pkg/results/reader.go

```diff
@@ -0,0 +1,227 @@
+	return &Reader{
+		Reader:  reader,
+		Version: version,
+	}
+}
+
+// NewReaderFromBytes is a helper constructor that will discover the version of the archive
+// and return a new Reader with the correct version already populated.
+func NewReaderFromBytes(data []byte) (*Reader, error) {
+	r := bytes.NewReader(data)
+	gzipReader, err := gzip.NewReader(r)
+	if err != nil {
+		return nil, err
+	}
+	version, err := DiscoverVersion(gzipReader)
+	if err != nil {
+		return nil, err
+	}
+	_, err = r.Seek(0, io.SeekStart)
+	if err != nil {
+		return nil, err
+	}
+	err = gzipReader.Reset(r)
+	if err != nil {
+		return nil, err
+	}
+	return &Reader{
+		Reader:  gzipReader,
+		Version: version,
+	}, nil
+}
+
+// DiscoverVersion takes a Sonobuoy archive stream and extracts just the
+// version of the archive.
+func DiscoverVersion(reader io.Reader) (string, error) {
+	r := &Reader{
+		Reader: reader,
+	}
+
+	conf := &config.Config{}
+
+	err := r.WalkFiles(func(path string, info os.FileInfo, err error) error {
+		return ExtractConfig(path, info, conf)
+	})
+	if err != nil {
+		return "", err
+	}
+	var version string
+	// Get rid of any of the extra version information that doesn't affect archive layout.
+	// Example: v0.10.0-a2b3d4
+	if strings.HasPrefix(conf.Version, versionEight) {
+		version = versionEight
+	} else if strings.HasPrefix(conf.Version, versionNine) {
+		version = versionNine
+	} else if strings.HasPrefix(conf.Version, versionTen) {
+		version = versionTen
+	} else {
+		return "", errors.New("cannot discover Sonobuoy archive version")
+	}
+	return version, nil
+}
+
+// tarFileInfo implements os.FileInfo and extends the Sys() method to
+// return a reader to a file in a tar archive.
+type tarFileInfo struct {
+	os.FileInfo
+	io.Reader
+}
+
+// Sys is going to be an io.Reader to a file in a tar archive.
+// This is how data is extracted from the archive.
+func (t *tarFileInfo) Sys() interface{} {
+	return t.Reader
+}
+
+// WalkFiles walks all of the files in the archive.
+func (r *Reader) WalkFiles(walkfn filepath.WalkFunc) error {
+	tr := tar.NewReader(r)
+	var err error
+	var header *tar.Header
+	for {
+		header, err = tr.Next()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return err
+		}
+		info := &tarFileInfo{
+			header.FileInfo(),
+			tr,
+		}
+		err = walkfn(filepath.Clean(header.Name), info, err)
+	}
+	return nil
+
+}
+
+// Functions to be used within a walkfn
+
+// ExtractIntoStruct takes a predicate function and some file information
+// and decodes the contents of the file that matches the predicate into the
+// interface passed in (generally a pointer to a struct/slice).
+func ExtractIntoStruct(predicate func(string) bool, path string, info os.FileInfo, object interface{}) error {
+	if predicate(path) {
+		reader, ok := info.Sys().(io.Reader)
+		if !ok {
+			return errors.New("info.Sys() is not a reader")
+		}
+		decoder := json.NewDecoder(reader)
+		err := decoder.Decode(object)
+		if err != nil {
+			return err
```

Comment:
might be worth wrapping this

Reply:
I've got a todo at the top to go through and wrap them all, I'll be refining and refactoring after a merge